### PR TITLE
Incorporate C++ Buffer management and Seal global threadpool fix from arrow

### PR DESCRIPTION
**Pull request by pcmoritz**
This includes https://github.com/apache/arrow/pull/1939 and gets rid of the revert of https://github.com/apache/arrow/pull/1807

It includes the following relevant arrow improvements:

- Use the automatic C++ PlasmaBuffer management for the Python wrappers (i.e. Release is called automatically)
- Removes the Plasma seal global threadpool which caused problems if multiple PlasmaClients were used in the same process in different threads

### CODE REVIEW

File: src/common/cmake/Common.cmake [hunk 1]
```diff
@@ -10,7 +10,12 @@ if(UNIX AND NOT APPLE)
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -rdynamic")
 endif()
 
-set(FLATBUFFERS_VERSION "1.7.1")
+# The following is needed because in CentOS, the lib directory is named lib64
+if(EXISTS "/etc/redhat-release" AND CMAKE_SIZEOF_VOID_P EQUAL 8)
```
Comment (robertnishihara) [inline src/common/cmake/Common.cmake]: Can we use this to get rid of https://github.com/ray-project/ray/blob/dad465a2bfcd355e8a04afbfa1fd84364051a79e/thirdparty/scripts/build_arrow.sh#L85-L90 and related lines?

Comment (robertnishihara) [reply src/common/cmake/Common.cmake]: Relatedly, is this the canonical way of checking for CentOS? https://askubuntu.com/questions/459402/how-to-know-if-the-running-platform-is-ubuntu-or-centos-with-help-of-a-bash-scri suggests `/etc/os-release`.

File: thirdparty/scripts/build_flatbuffers.sh [hunk 2]
```diff
@@ -7,7 +7,7 @@ set -e
 
 TP_DIR=$(cd "$(dirname "${BASH_SOURCE:-$0}")"; pwd)/../
 
-FLATBUFFERS_VERSION=1.7.1
```
Comment (robertnishihara) [inline thirdparty/scripts/build_flatbuffers.sh]: Wait, are we building flatbuffers twice? (excluding the time that Arrow builds it)

Why do we have this script in addition to the `ExternalProject_Add(flatbuffers_ep` in `Common.cmake`?

**Comment by robertnishihara** [comment]
Can you make the PR title and description more informative? E.g., what changes are we getting from Arrow?

File: src/common/cmake/Common.cmake (hunk 1 above)
Comment (pcmoritz) [reply src/common/cmake/Common.cmake]: I took it from https://htcondor-wiki.cs.wisc.edu/index.cgi/fileview?f=build/cmake/FindLinuxPlatform.cmake&v=4592599fecc08e5588c4244e2b0ceb7d32363a56, it's officially documented by Red Hat, so should be fine.

File: thirdparty/scripts/build_flatbuffers.sh (hunk 2 above)
Comment (pcmoritz) [reply thirdparty/scripts/build_flatbuffers.sh]: Yeah, I think we can figure this out when we get rid of the old ray code (and get rid of Common.cmake then).

File: src/common/cmake/Common.cmake (hunk 1 above)
Comment (pcmoritz) [reply src/common/cmake/Common.cmake]: Getting rid of the copy in arrow will involve changing FindArrow, which is part of arrow.

**Review by robertnishihara — APPROVED** [review]

**Pull request merged**